File: package/go/src/server_log_analysis.go
package src

// LogEntry represents a server log entry
type LogEntry struct {
	IPAddress  string // IP address
	Request    string // Request string
	StatusCode int    // HTTP status code
}

// FilterSuccessfulRequests filters logs to return only requests with status code 200.
//
// Parameters:
//   - logs: slice of LogEntry structs
//
// Returns:
//   - slice of logs with status code 200
func FilterSuccessfulRequests(logs []LogEntry) []LogEntry {
	result := make([]LogEntry, 0)
	for _, log := range logs {
		if log.StatusCode == 200 {
			result = append(result, log)
		}
	}
	return result
}

// CountRequestsByIP counts the number of requests per IP address.
//
// Parameters:
//   - logs: slice of LogEntry structs
//
// Returns:
//   - map of IP address to request count
func CountRequestsByIP(logs []LogEntry) map[string]int {
	ipCounts := make(map[string]int)
	for _, log := range logs {
		ipCounts[log.IPAddress]++
	}
	return ipCounts
}

File: package/rust/src/server_log_analysis.rs
use std::collections::HashMap;

pub fn filter_successful_requests(logs: &[(String, String, i32)]) -> Vec<(String, String, i32)> {
    logs.iter()
        .filter(|(_, _, status_code)| *status_code == 200)
        .cloned()
        .collect()
}

pub fn count_requests_by_ip(logs: &[(String, String, i32)]) -> HashMap<String, i32> {
    let mut ip_counts = HashMap::new();

    for (ip, _, _) in logs {
        *ip_counts.entry(ip.clone()).or_insert(0) += 1;
    }

    ip_counts
}
